File: run-tests.py
#!/usr/bin/env python3
"""Run backend and frontend tests inside Docker containers.

- Starts infrastructure with ENVIRONMENT=test
- Runs layer2ledger unit tests before apihandler/dbwriter (avoids DB/Redis lock contention)
- Starts application services, then runs integration test containers (including wallet vitest)
- Leaves a running healthy bitcoin-core container untouched
"""

from __future__ import annotations

import json
import os
import shutil
import subprocess
import sys
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parent

COMPOSE_FILES = ("docker-compose.yml", "docker-compose.test.yml")

INFRA_SERVICES = (
    "layer2ledger-postgres",
    "layer2ledger-redis",
    "layer2ledger-mongodb",
)

APP_SERVICES = (
    "layer2ledgerapihandler",
    "layer2ledgerdbwriter",
    "layer2ledgeroauthmanager",
    "layer2bridge",
)

TEST_SERVICES = (
    "test-layer2ledger",
    "test-layer2bridge",
    "test-bitcoin-core-rpc",
    "test-layer2ledgeroauthmanager",
    "test-layer2ledger-seed",
    "test-wallet-web",
)

UNIT_TEST_SERVICES = ("test-layer2ledger",)

INTEGRATION_TEST_SERVICES = (
    "test-layer2bridge",
    "test-bitcoin-core-rpc",
    "test-layer2ledgeroauthmanager",
    "test-layer2ledger-seed",
    "test-wallet-web",
)

# Long-running services in the "test" compose profile. `compose run` starts these as
# dependencies but does not stop them when the one-shot test container exits.
PROFILE_BACKGROUND_SERVICES = ("layer2ledger-testhelper",)

SETUP_UV_IMAGE = os.environ.get(
    "SETUP_UV_IMAGE", "ghcr.io/astral-sh/uv:python3.12-bookworm"
)

CONFIG_MARKER = ROOT / ".config/test/layer2ledgerbatched-common-config.json"
ENV_TEST = ROOT / "backend/layer2ledgerbatched/.env.test"


def log(message: str) -> None:
    print(f"==> {message}", flush=True)


def load_env_file(path: Path) -> dict[str, str]:
    values: dict[str, str] = {}
    for line in path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, value = line.split("=", 1)
        values[key.strip()] = value.strip()
    return values


class DockerComposeTestRunner:
    def __init__(self, root: Path) -> None:
        self.root = root
        self.env = os.environ.copy()
        self.env["ENVIRONMENT"] = "test"
        self.compose = ["docker", "compose", "--progress", "quiet"]
        for compose_file in COMPOSE_FILES:
            self.compose.extend(["-f", compose_file])

    def run(
        self,
        args: list[str],
        *,
        check: bool = True,
        capture_output: bool = False,
    ) -> subprocess.CompletedProcess[str]:
        command = [*self.compose, *args]
        return subprocess.run(
            command,
            cwd=self.root,
            env=self.env,
            check=check,
            text=True,
            capture_output=capture_output,
        )

    def run_quiet(self, args: list[str]) -> bool:
        return self.run(args, check=False).returncode == 0

    def docker_run(self, args: list[str], *, check: bool = True) -> subprocess.CompletedProcess[str]:
        return subprocess.run(args, cwd=self.root, check=check, text=True)

    def run_setup_scripts(self, *script_args: str) -> None:
        setup_dir = self.root / "backend/setup_scripts"
        env = self.env.copy()
        env["OPENL2_CONFIG_PATH"] = str(self.root / ".config")

        if shutil.which("uv"):
            subprocess.run(
                ["uv", "run", *script_args],
                cwd=setup_dir,
                env=env,
                check=True,
            )
            return

        log(f"uv not found on host — running setup scripts in Docker ({SETUP_UV_IMAGE})...")
        self.docker_run(
            [
                "docker",
                "run",
                "--rm",
                "-v",
                f"{self.root}:/workspace",
                "-w",
                "/workspace/backend/setup_scripts",
                "-e",
                f"OPENL2_CONFIG_PATH={env['OPENL2_CONFIG_PATH']}",
                SETUP_UV_IMAGE,
                "sh",
                "-ec",
                'uv sync --frozen && exec uv run "$@"',
                "_",
                *script_args,
            ]
        )

    def ensure_test_config(self) -> None:
        if CONFIG_MARKER.is_file():
            log(f"Test config found at {CONFIG_MARKER.relative_to(self.root)}")
            return

        log("Generating test config (.config/test/)...")
        self.run_setup_scripts(
            "src/main.py",
            "-env",
            "test",
            "-containered",
            "true",
            "-generate-keys",
            "-generate-oauth-config",
        )

    def ensure_test_postgres_database(self) -> None:
        db_name = load_env_file(ENV_TEST)["POSTGRES_DB"]
        log(f"Ensuring PostgreSQL database exists: {db_name}")

        result = self.run(
            [
                "exec",
                "-T",
                "layer2ledger-postgres",
                "psql",
                "-U",
                "postgres",
                "-tc",
                f"SELECT 1 FROM pg_database WHERE datname = '{db_name}'",
            ],
            check=True,
            capture_output=True,
        )
        if "1" in result.stdout:
            log(f"PostgreSQL database {db_name} already exists")
            return

        self.run(
            [
                "exec",
                "-T",
                "layer2ledger-postgres",
                "psql",
                "-U",
                "postgres",
                "-c",
                f"CREATE DATABASE {db_name};",
            ],
            check=True,
        )
        log(f"Created PostgreSQL database {db_name}")

    def container_id(self, service: str) -> str:
        result = self.run(["ps", "-q", service], check=False, capture_output=True)
        return result.stdout.strip()

    def container_health(self, container_id: str) -> str:
        if not container_id:
            return "missing"

        result = subprocess.run(
            ["docker", "inspect", container_id],
            capture_output=True,
            text=True,
            check=False,
        )
        if result.returncode != 0:
            return "unknown"

        state = json.loads(result.stdout)[0]["State"]
        health = state.get("Health")
        if health and health.get("Status"):
            return health["Status"]
        return state.get("Status", "unknown")

    def wait_for_healthy(self, service: str, timeout_sec: int = 180) -> None:
        waited = 0
        while waited < timeout_sec:
            health = self.container_health(self.container_id(service))
            if health == "healthy":
                log(f"{service} is healthy")
                return
            time.sleep(2)
            waited += 2

        log(f"ERROR: timed out waiting for {service} to become healthy")
        self.run(["ps"], check=False)
        raise SystemExit(1)

    def ensure_bitcoin_core(self) -> None:
        container_id = self.container_id("bitcoin-core")
        if container_id:
            health = self.container_health(container_id)
            if health != "healthy":
                log(f"ERROR: bitcoin-core is running but not healthy (status: {health}).")
                log("Fix bitcoin-core manually. This script will not restart or recreate it.")
                raise SystemExit(1)
            log("bitcoin-core already running and healthy — leaving unchanged")
            return

        log("bitcoin-core is not running — starting it (first-time only)...")
        self.run(["up", "-d", "bitcoin-core"], check=True)
        self.wait_for_healthy("bitcoin-core", timeout_sec=600)

    def start_infra_services(self) -> None:
        log("Starting infrastructure for test (ENVIRONMENT=test)...")
        self.run(
            ["up", "-d", "--build", "--force-recreate", *INFRA_SERVICES],
            check=True,
        )
        for service in INFRA_SERVICES:
            self.wait_for_healthy(service)
        self.ensure_test_postgres_database()

    def stop_app_services(self) -> None:
        log("Stopping application services so pytest can use Postgres/Redis exclusively...")
        self.run_quiet(["stop", *APP_SERVICES])

    def start_app_services(self) -> None:
        log("Starting backend application services for test (ENVIRONMENT=test)...")
        self.run(
            ["up", "-d", "--build", "--force-recreate", *APP_SERVICES],
            check=True,
        )
        for service in APP_SERVICES:
            self.wait_for_healthy(service)

    def ensure_testhelper(self) -> None:
        """Rebuild testhelper so it always runs the current seed.py (not a stale image).

        `compose run` reuses a healthy layer2ledger-testhelper left over from
        `make backend-test` or a prior `make test` without rebuilding it.
        """
        log("Building and starting layer2ledger-testhelper (force-recreate)...")
        self.run(
            [
                "--profile",
                "test",
                "up",
                "-d",
                "--build",
                "--force-recreate",
                "layer2ledger-testhelper",
            ],
            check=True,
        )
        self.wait_for_healthy("layer2ledger-testhelper")

    def run_test_services(self, services: tuple[str, ...]) -> bool:
        failed = False
        for test_service in services:
            log(f"Running {test_service}...")
            if self.run_quiet(
                ["--profile", "test", "run", "--rm", "--build", test_service]
            ):
                log(f"PASSED: {test_service}")
            else:
                log(f"FAILED: {test_service}")
                failed = True
        return failed

    def stop_test_profile_services(self) -> None:
        log("Stopping test-profile background services...")
        self.run_quiet(["--profile", "test", "stop", *PROFILE_BACKGROUND_SERVICES])

    def cleanup_test_containers(self) -> None:
        self.stop_test_profile_services()
        log("Removing stopped test containers (if any)...")
        self.run_quiet(["--profile", "test", "rm", "-sf", *TEST_SERVICES])

    def main(self) -> int:
        self.ensure_test_config()
        self.ensure_bitcoin_core()
        self.start_infra_services()
        self.stop_app_services()

        exit_code = 0

        log("Running layer2ledger unit tests (no live apihandler/dbwriter)...")
        if self.run_test_services(UNIT_TEST_SERVICES):
            exit_code = 1

        self.start_app_services()

        self.ensure_testhelper()

        log("Running integration test containers...")
        if self.run_test_services(INTEGRATION_TEST_SERVICES):
            exit_code = 1

        self.cleanup_test_containers()

        if exit_code == 0:
            log("All test containers passed")
        else:
            log("One or more test containers failed")

        return exit_code


def main() -> None:
    os.chdir(ROOT)
    runner = DockerComposeTestRunner(ROOT)
    raise SystemExit(runner.main())


if __name__ == "__main__":
    main()

File: run-tests.ts
#!/usr/bin/env bun
/**
 * Run backend and frontend tests inside Docker containers.
 *
 * - Starts infrastructure with ENVIRONMENT=test
 * - Runs layer2ledger unit tests before apihandler/dbwriter (avoids DB/Redis lock contention)
 * - Starts application services, then runs integration test containers (including wallet vitest)
 * - Leaves a running healthy bitcoin-core container untouched
 */

import { existsSync, readFileSync } from 'fs';
import { join } from 'path';

const ROOT = import.meta.dir;

const COMPOSE_FILES = ['docker-compose.yml', 'docker-compose.test.yml'] as const;

const INFRA_SERVICES = [
  'layer2ledger-postgres',
  'layer2ledger-redis',
  'layer2ledger-mongodb',
] as const;

const APP_SERVICES = [
  'layer2ledgerapihandler',
  'layer2ledgerdbwriter',
  'layer2ledgeroauthmanager',
  'layer2bridge',
] as const;

const TEST_SERVICES = [
  'test-layer2ledger',
  'test-layer2bridge',
  'test-bitcoin-core-rpc',
  'test-layer2ledgeroauthmanager',
  'test-layer2ledger-seed',
  'test-wallet-web',
] as const;

const UNIT_TEST_SERVICES = ['test-layer2ledger'] as const;

const INTEGRATION_TEST_SERVICES = [
  'test-layer2bridge',
  'test-bitcoin-core-rpc',
  'test-layer2ledgeroauthmanager',
  'test-layer2ledger-seed',
  'test-wallet-web',
] as const;

/** Long-running services in the "test" compose profile. */
const PROFILE_BACKGROUND_SERVICES = ['layer2ledger-testhelper'] as const;

const SETUP_BUN_IMAGE = process.env.SETUP_BUN_IMAGE ?? 'oven/bun:1.3';

const CONFIG_MARKER = join(ROOT, '.config/test/layer2ledger-common-config.json');
const ENV_TEST = join(ROOT, 'backend/layer2ledger/.env.test');

function log(message: string): void {
  console.log(`==> ${message}`);
}

function loadEnvFile(path: string): Record<string, string> {
  const values: Record<string, string> = {};
  for (const line of readFileSync(path, 'utf-8').split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#') || !trimmed.includes('=')) {
      continue;
    }
    const separatorIndex = trimmed.indexOf('=');
    const key = trimmed.slice(0, separatorIndex).trim();
    const value = trimmed.slice(separatorIndex + 1).trim();
    values[key] = value;
  }
  return values;
}

interface InspectState {
  Status?: string;
  Health?: { Status?: string };
}

class DockerComposeTestRunner {
  private readonly env: Record<string, string>;
  private readonly compose: string[];

  constructor(private readonly root: string) {
    this.env = { ...process.env, ENVIRONMENT: 'test' } as Record<string, string>;
    this.compose = ['docker', 'compose', '--progress', 'quiet'];
    for (const composeFile of COMPOSE_FILES) {
      this.compose.push('-f', composeFile);
    }
  }

  private async run(
    args: string[],
    options?: { check?: boolean; captureOutput?: boolean },
  ): Promise<{ exitCode: number; stdout: string; stderr: string }> {
    const check = options?.check ?? true;
    const captureOutput = options?.captureOutput ?? false;
    const command = [...this.compose, ...args];

    const proc = Bun.spawn(command, {
      cwd: this.root,
      env: this.env,
      stdout: captureOutput ? 'pipe' : 'inherit',
      stderr: captureOutput ? 'pipe' : 'inherit',
    });

    const stdout = captureOutput ? await new Response(proc.stdout).text() : '';
    const stderr = captureOutput ? await new Response(proc.stderr).text() : '';
    const exitCode = await proc.exited;

    if (check && exitCode !== 0) {
      throw new Error(`Command failed (${exitCode}): ${command.join(' ')}`);
    }

    return { exitCode, stdout, stderr };
  }

  private async runQuiet(args: string[]): Promise<boolean> {
    const result = await this.run(args, { check: false });
    return result.exitCode === 0;
  }

  private async dockerRun(args: string[], check = true): Promise<void> {
    const proc = Bun.spawn(args, {
      cwd: this.root,
      stdout: 'inherit',
      stderr: 'inherit',
    });
    const exitCode = await proc.exited;
    if (check && exitCode !== 0) {
      throw new Error(`Command failed (${exitCode}): ${args.join(' ')}`);
    }
  }

  private async runSetupScripts(...scriptArgs: string[]): Promise<void> {
    const setupDir = join(this.root, 'backend/setup_scripts');
    const env = {
      ...this.env,
      OPENL2_CONFIG_PATH: join(this.root, '.config'),
    };

    if (Bun.which('bun')) {
      const proc = Bun.spawn(['bun', 'run', 'src/main.ts', ...scriptArgs], {
        cwd: setupDir,
        env,
        stdout: 'inherit',
        stderr: 'inherit',
      });
      const exitCode = await proc.exited;
      if (exitCode !== 0) {
        throw new Error(`Setup scripts failed with exit code ${exitCode}`);
      }
      return;
    }

    log(`bun not found on host — running setup scripts in Docker (${SETUP_BUN_IMAGE})...`);
    await this.dockerRun([
      'docker',
      'run',
      '--rm',
      '-v',
      `${this.root}:/workspace`,
      '-w',
      '/workspace/backend/setup_scripts',
      '-e',
      `OPENL2_CONFIG_PATH=${env.OPENL2_CONFIG_PATH}`,
      SETUP_BUN_IMAGE,
      'bun',
      'run',
      'src/main.ts',
      ...scriptArgs,
    ]);
  }

  private async ensureTestConfig(): Promise<void> {
    if (existsSync(CONFIG_MARKER)) {
      log(`Test config found at ${CONFIG_MARKER.slice(this.root.length + 1)}`);
      return;
    }

    log('Generating test config (.config/test/)...');
    await this.runSetupScripts(
      '-env',
      'test',
      '-containered',
      'true',
      '-generate-keys',
      '-generate-oauth-config',
    );
  }

  private async ensureTestPostgresDatabase(): Promise<void> {
    const dbName = loadEnvFile(ENV_TEST).POSTGRES_DB;
    if (!dbName) {
      throw new Error(`POSTGRES_DB missing from ${ENV_TEST}`);
    }
    log(`Ensuring PostgreSQL database exists: ${dbName}`);

    const result = await this.run(
      [
        'exec',
        '-T',
        'layer2ledger-postgres',
        'psql',
        '-U',
        'postgres',
        '-tc',
        `SELECT 1 FROM pg_database WHERE datname = '${dbName}'`,
      ],
      { check: true, captureOutput: true },
    );

    if (result.stdout.includes('1')) {
      log(`PostgreSQL database ${dbName} already exists`);
      return;
    }

    await this.run(
      [
        'exec',
        '-T',
        'layer2ledger-postgres',
        'psql',
        '-U',
        'postgres',
        '-c',
        `CREATE DATABASE ${dbName};`,
      ],
      { check: true },
    );
    log(`Created PostgreSQL database ${dbName}`);
  }

  private async containerId(service: string): Promise<string> {
    const result = await this.run(['ps', '-q', service], {
      check: false,
      captureOutput: true,
    });
    return result.stdout.trim();
  }

  private async containerHealth(containerId: string): Promise<string> {
    if (!containerId) {
      return 'missing';
    }

    const proc = Bun.spawn(['docker', 'inspect', containerId], {
      stdout: 'pipe',
      stderr: 'pipe',
    });
    const stdout = await new Response(proc.stdout).text();
    const exitCode = await proc.exited;
    if (exitCode !== 0) {
      return 'unknown';
    }

    const inspected = JSON.parse(stdout) as Array<{ State: InspectState }>;
    const state = inspected[0]?.State;
    if (!state) {
      return 'unknown';
    }
    if (state.Health?.Status) {
      return state.Health.Status;
    }
    return state.Status ?? 'unknown';
  }

  private async waitForHealthy(service: string, timeoutSec = 180): Promise<void> {
    let waited = 0;
    while (waited < timeoutSec) {
      const health = await this.containerHealth(await this.containerId(service));
      if (health === 'healthy') {
        log(`${service} is healthy`);
        return;
      }
      await Bun.sleep(2000);
      waited += 2;
    }

    log(`ERROR: timed out waiting for ${service} to become healthy`);
    await this.run(['ps'], { check: false });
    throw new Error(`Timed out waiting for ${service}`);
  }

  private async ensureBitcoinCore(): Promise<void> {
    const id = await this.containerId('bitcoin-core');
    if (id) {
      const health = await this.containerHealth(id);
      if (health !== 'healthy') {
        log(`ERROR: bitcoin-core is running but not healthy (status: ${health}).`);
        log('Fix bitcoin-core manually. This script will not restart or recreate it.');
        throw new Error('bitcoin-core unhealthy');
      }
      log('bitcoin-core already running and healthy — leaving unchanged');
      return;
    }

    log('bitcoin-core is not running — starting it (first-time only)...');
    await this.run(['up', '-d', 'bitcoin-core'], { check: true });
    await this.waitForHealthy('bitcoin-core', 600);
  }

  private async startInfraServices(): Promise<void> {
    log('Starting infrastructure for test (ENVIRONMENT=test)...');
    await this.run(['up', '-d', '--build', '--force-recreate', ...INFRA_SERVICES], {
      check: true,
    });
    for (const service of INFRA_SERVICES) {
      await this.waitForHealthy(service);
    }
    await this.ensureTestPostgresDatabase();
  }

  private async stopAppServices(): Promise<void> {
    log('Stopping application services so unit tests can use Postgres/Redis exclusively...');
    await this.runQuiet(['stop', ...APP_SERVICES]);
  }

  private async startAppServices(): Promise<void> {
    log('Starting backend application services for test (ENVIRONMENT=test)...');
    await this.run(['up', '-d', '--build', '--force-recreate', ...APP_SERVICES], {
      check: true,
    });
    for (const service of APP_SERVICES) {
      await this.waitForHealthy(service);
    }
  }

  private async ensureTesthelper(): Promise<void> {
    log('Building and starting layer2ledger-testhelper (force-recreate)...');
    await this.run(
      [
        '--profile',
        'test',
        'up',
        '-d',
        '--build',
        '--force-recreate',
        'layer2ledger-testhelper',
      ],
      { check: true },
    );
    await this.waitForHealthy('layer2ledger-testhelper');
  }

  private async runTestServices(services: readonly string[]): Promise<boolean> {
    let failed = false;
    for (const testService of services) {
      log(`Running ${testService}...`);
      if (await this.runQuiet(['--profile', 'test', 'run', '--rm', '--build', testService])) {
        log(`PASSED: ${testService}`);
      } else {
        log(`FAILED: ${testService}`);
        failed = true;
      }
    }
    return failed;
  }

  private async stopTestProfileServices(): Promise<void> {
    log('Stopping test-profile background services...');
    await this.runQuiet(['--profile', 'test', 'stop', ...PROFILE_BACKGROUND_SERVICES]);
  }

  private async cleanupTestContainers(): Promise<void> {
    await this.stopTestProfileServices();
    log('Removing stopped test containers (if any)...');
    await this.runQuiet(['--profile', 'test', 'rm', '-sf', ...TEST_SERVICES]);
  }

  async main(): Promise<number> {
    await this.ensureTestConfig();
    await this.ensureBitcoinCore();
    await this.startInfraServices();
    await this.stopAppServices();

    let exitCode = 0;

    log('Running layer2ledger unit tests (no live apihandler/dbwriter)...');
    if (await this.runTestServices(UNIT_TEST_SERVICES)) {
      exitCode = 1;
    }

    await this.startAppServices();
    await this.ensureTesthelper();

    log('Running integration test containers...');
    if (await this.runTestServices(INTEGRATION_TEST_SERVICES)) {
      exitCode = 1;
    }

    await this.cleanupTestContainers();

    if (exitCode === 0) {
      log('All test containers passed');
    } else {
      log('One or more test containers failed');
    }

    return exitCode;
  }
}

const runner = new DockerComposeTestRunner(ROOT);
process.exit(await runner.main());
